src/main/java/org/prebid/server/auction/ExchangeService.java

```diff
@@ -672,12 +672,88 @@ private static ExtPriceGranularity parsePriceGranularity(JsonNode priceGranulari
      * <p>
      * Returns input argument as the result if no errors found or create new {@link BidderSeatBid} otherwise.
      */
-    private BidderSeatBid validateAndUpdateResponse(BidderSeatBid bidderSeatBid) {
+    private BidderSeatBid validateAndUpdateResponse(BidderSeatBid bidderSeatBid, List<String> requestCurrencies) {
+        final List<String> effectiveRequestCurrencies = requestCurrencies.isEmpty()
+                ? Collections.singletonList(DEFAULT_CURRENCY) : requestCurrencies;
+
         final List<BidderBid> bids = bidderSeatBid.getBids();
 
         final List<BidderBid> validBids = new ArrayList<>(bids.size());
         final List<BidderError> errors = new ArrayList<>(bidderSeatBid.getErrors());
 
+        if (isValidCurFromResponse(effectiveRequestCurrencies, bids, errors)) {
+            validateResponseBids(bids, validBids, errors);
+        }
+
+        return validBids.size() == bids.size()
+                ? bidderSeatBid
+                : BidderSeatBid.of(validBids, bidderSeatBid.getHttpCalls(), errors);
+    }
+
+    private boolean isValidCurFromResponse(List<String> requestCurrencies, List<BidderBid> bids,
+                                           List<BidderError> errors) {
+        if (bids.size() > 0) {
+            //assume that currencies are the same among all bids
+            String bidsCurrency = nonNullCurrencyOtherwiseDefault(bids.get(0).getBidCurrency());
+
+            boolean sameCurrenciesAcrossAllBids = isValidCurrencyPresent(bids);
+
+            if (!sameCurrenciesAcrossAllBids) {
+                errors.add(BidderError.generic("Bid currencies mismatch found. "
+                        + "Expected all bids to have the same currencies."));
+                return false;
+            }
+
+            boolean validCurrencyPresent = isValidCurrencyPresent(requestCurrencies, bidsCurrency);
+
+            if (!validCurrencyPresent) {
+                final String requestCursFormatted = requestCurrencies.stream()
+                        .collect(Collectors.joining(","));
+                errors.add(BidderError.generic(String.format(
+                        "Bid currency is not allowed. Was %s, wants: [%s]",
+                        requestCursFormatted, bidsCurrency)));
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean isValidCurrencyPresent(List<BidderBid> bids) {
+        boolean sameCurrenciesAcrossAllBids = true;
```

Comment:
This method could be simplified using streams: `bids.stream().map(map to currency).distinct().count() ==1`